okhttp/src/main/java/com/squareup/okhttp/ConnectionPool.java

```diff
@@ -166,4 +210,64 @@ public void evictAll() {
       Util.closeQuietly(connection.getSocket());
     }
   }
+
+  /**
+   * Performs maintenance on this pool, evicting connections that have expired.
+   *
+   * <p>Returns the duration in nanos to sleep until the next scheduled call to this method.
+   * Returns -1 if no further cleanups are required.
+   */
+  long cleanup(long now) {
+    int inUseConnectionCount = 0;
+    long nanosUntilNextCleanup = -1L;
+    RealConnection connectionToEvict = null;
+
+    // Find either a connection to evict, or the time that the next eviction is due.
+    synchronized (this) {
+      for (Iterator<RealConnection> i = connections.iterator(); i.hasNext(); ) {
+        RealConnection connection = i.next();
+
+        // If the connection is in use, keep searching.
+        if (connection.allocationCount > 0) {
+          inUseConnectionCount++;
+          nanosUntilNextCleanup = keepAliveDurationNs;
+          continue;
+        }
+
+        // If the connection is ready to be evicted, we're done.
+        long evictAtNanos = connection.idleAtNanos + keepAliveDurationNs;
+        long nanosUntilEviction = evictAtNanos - now;
+        if (nanosUntilEviction <= 0) {
+          connection.noNewStreams = true;
+          connectionToEvict = connection;
+          i.remove();
+          break;
+        }
+
+        // Is this the next connection to evict?
+        if (nanosUntilNextCleanup == -1L || nanosUntilNextCleanup > nanosUntilEviction) {
+          nanosUntilNextCleanup = nanosUntilEviction;
+        }
+      }
+    }
+
+    if (connectionToEvict != null) {
+      Util.closeQuietly(connectionToEvict.getSocket());
+      // Cleanup again immediately.
+      return 0;
+    } else if (nanosUntilNextCleanup != -1) {
+      // A connection will be ready to evict soon.
+      return nanosUntilNextCleanup;
+    } else if (inUseConnectionCount != 0) {
+      // All connections are in use. It'll be at least the keep alive duration 'til we run again.
+      return keepAliveDurationNs;
+    } else {
+      // No connections, idle or in use.
+      return -1;
+    }
+  }
+
+  void setCleanupRunnableForTest(Runnable cleanupRunnable) {
```

Comment:
You could use a packaged-scoped third constructor as an alternative. I don't really care either way.